tls/s2n_server_extensions.c

```diff
@@ -13,176 +13,56 @@
-
+#include "tls/s2n_connection.h"
 #include "stuffer/s2n_stuffer.h"
-
 #include "utils/s2n_safety.h"
-#include "utils/s2n_blob.h"
-
-/* compute size server extensions send requires */
-int s2n_server_extensions_send_size(struct s2n_connection *conn)
-{
-    int total_size = 0;
-    const bool is_tls13_conn = conn->actual_protocol_version == S2N_TLS13;
-
-    if (is_tls13_conn) {
-        GUARD_UINT16_AND_INCREMENT(s2n_extensions_server_supported_versions_size(conn), total_size);
-        GUARD_UINT16_AND_INCREMENT(s2n_extensions_server_key_share_send_size(conn), total_size);
-        GUARD_UINT16_AND_INCREMENT(s2n_extensions_cookie_size(conn), total_size);
-        return total_size;
-    }
 
-    GUARD_UINT16_AND_INCREMENT(s2n_server_extensions_server_name_send_size(conn), total_size);
-    GUARD_UINT16_AND_INCREMENT(s2n_server_extensions_alpn_send_size(conn), total_size);
-    GUARD_UINT16_AND_INCREMENT(s2n_server_renegotiation_info_ext_size(conn), total_size);
-    GUARD_UINT16_AND_INCREMENT(s2n_server_ecc_point_format_extension_size(conn), total_size);
-    GUARD_UINT16_AND_INCREMENT(s2n_server_extensions_max_fragment_length_send_size(conn), total_size);
-    GUARD_UINT16_AND_INCREMENT(s2n_server_session_ticket_ext_size(conn), total_size);
-    GUARD_UINT16_AND_INCREMENT(s2n_server_extensions_status_request_send_size(conn), total_size);
-    GUARD_UINT16_AND_INCREMENT(s2n_server_extensions_sct_list_send_size(conn), total_size);
-
-    return total_size;
-}
+/* An empty list will just contain the uint16_t list size */
+#define S2N_EMPTY_EXTENSION_LIST_SIZE sizeof(uint16_t)
 
 int s2n_server_extensions_send(struct s2n_connection *conn, struct s2n_stuffer *out)
 {
-    int total_size = s2n_server_extensions_send_size(conn);
+    uint32_t data_available_before_extensions = s2n_stuffer_data_available(out);
 
-    if (total_size == 0) {
-        return 0;
+    if (conn->actual_protocol_version >= S2N_TLS13) {
+        GUARD(s2n_extension_list_send(S2N_EXTENSION_LIST_SERVER_HELLO_TLS13, conn, out));
+    } else {
+        GUARD(s2n_extension_list_send(S2N_EXTENSION_LIST_SERVER_HELLO_DEFAULT, conn, out));
     }
-    GUARD_UINT16(total_size);
-
-    GUARD(s2n_stuffer_write_uint16(out, total_size));
-
-    const bool is_tls13_conn = conn->actual_protocol_version == S2N_TLS13;
 
-    /* TLS 1.3 ServerHello extensions */
-    if (is_tls13_conn) {
-        /* Write supported versions extension */
-        GUARD(s2n_extensions_server_supported_versions_send(conn, out));
-        /* Write key share extension */
-        GUARD(s2n_extensions_server_key_share_send(conn, out));
-        /* Write cookie extension */
-        GUARD(s2n_extensions_cookie_send(conn, out));
-
-        return 0;
+    /* The ServerHello extension list size (uint16_t) is NOT written if the list is empty.
+     * This is to support older clients written before extensions existed that might fail
+     * on any unexpected bytes at the end of the ServerHello.
```

Comment:
Is there a test that codifies this behavior and specifically tests the ServerHello bytes?

Reply:
[This one](https://github.com/awslabs/s2n/blob/6ab5166f19d16e02757c8a4a7672d44a666dc513/tests/unit/s2n_server_extensions_test.c#L75-L86) already existed. It checks that send writes nothing for an empty list.